src/ResourceManager/Resources/Commands.Resources/Providers/GetAzureLocationCmdlet.cs

```diff
@@ -0,0 +1,107 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// ----------------------------------------------------------------------------------
+
+namespace Microsoft.Azure.Commands.Providers
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Management.Automation;
+    using Microsoft.Azure.Commands.ResourceManager.Cmdlets.Extensions;
+    using Microsoft.Azure.Commands.Resources.Models;
+    using Microsoft.Azure.Management.Resources.Models;
+    using Microsoft.Azure.Subscriptions.Models;
+    using Microsoft.WindowsAzure.Commands.Utilities.Common;
+
+    /// <summary>
+    /// Get all locations with the supported providers.
+    /// </summary>
+    [Cmdlet(VerbsCommon.Get, "AzureRmLocation"), OutputType(typeof(List<PSResourceProviderLocation>))]
+    public class GetAzureLocationCmdlet : ResourcesBaseCmdlet
+    {
+        /// <summary>
+        /// Executes the cmdlet
+        /// </summary>
+        public override void ExecuteCmdlet()
+        {
+            WriteWarning("The output object type of this cmdlet will be modified in a future release.");
+
+            var allLocations = this.ResourcesClient.ListLocations(DefaultContext.Subscription.Id.ToString());
+            var providers = this.ResourcesClient.ListResourceProviders(providerName: null, listAvailable: true);
+
+            var providerLocations = ConstructResourceProviderLocations(allLocations, providers);
+
+            this.WriteObject(providerLocations, enumerateCollection: true);   
+        }
+
+        private List<PSResourceProviderLocation> ConstructResourceProviderLocations(List<Location> locations, List<Provider> providers)
+        {
+            var locationProviderMap = GetLocationProviderMap(providers);
+
+            // Join locations and locationProviderMap on DisplayName
+            // locations: list of [Name, DisplayName, Id, ...]
+            // locationProviderMap: list of [Key: DisplayName, Value: Providers]
+            var joinResult = locations.CoalesceEnumerable().Join(
+                locationProviderMap.CoalesceEnumerable(),
+                location => location.DisplayName,
+                mapEntry => mapEntry.Key,
+                (location, mapEntry) => new PSResourceProviderLocation
+                    {
+                        Location = location.Name,
+                        DisplayName = location.DisplayName,
+                        Providers = mapEntry.Value
+                    },
+                StringComparer.InvariantCultureIgnoreCase);
+
+            return joinResult.ToList();
+        }
+
+        private Dictionary<string, List<string>> GetLocationProviderMap(List<Provider> providers)
+        {
+            var locationMap = new Dictionary<string, List<string>>();
```

Comment:
Consider using a case-insensitive StringComparer in the dictionary constructor.  This will prevent multiple keys due to differences in capitalization